File: jasmine/src/api/site.ts
import { query, ISucceed } from "./graphql";

export interface IRoute {
  path: string;
  name: string;
  children: IRoute[];
}

export const routes = async (): Promise<IRoute[]> => {
  const res = await query<{
    indexRoute: IRoute[];
  }>(
    `
query call{
  indexRoute{
    path, name,
    children {
      path, name, children{
        path, name, children{
          path, name
        }
      }
    }
  }
}
`,
    {}
  );
  return res.indexRoute;
};

interface IGab {
  code: string;
  name: string;
}
interface IIcp {
  code: string;
}
interface IAuthor {
  name: string;
  email: string;
}
export interface ILayout {
  favicon: string;
  title: string;
  subhead: string;
  description: string;
  author?: IAuthor;
  keywords: string[];
  copyright: string;
  locale: string;
  languages: string[];
  icp?: IIcp;
  gab?: IGab;
}

export interface ILayoutResponse {
  apiVersion: string;
  layout: ILayout;
}
export const fetch_layout = async (): Promise<ILayoutResponse> => {
  const res = await query<ILayoutResponse>(
    `
query call{
  apiVersion
  layout{
    favicon, title, subhead, description, keywords, copyright, 
    locale, languages, 
    author{name, email}, 
    icp{code}, 
    gab{code, name},
  }
}
`,
    {}
  );
  return res;
};

export interface IPostgreSqlStatus {
  version: string;
  migrations: string[];
}
export interface IRedisStatus {
  nodes: string[];
}

export interface ISystemStatusResponse {
  postgresql: IPostgreSqlStatus;
  redis: IRedisStatus;
}

export const system_status = async (): Promise<ISystemStatusResponse> => {
  const res = await query<{
    postgresqlStatus: IPostgreSqlStatus;
    redisStatus: IRedisStatus;
  }>(
    `
query call{
  postgresqlStatus{
    version, migrations
  }
  redisStatus{
    nodes
  }
}
`,
    {}
  );
  return { postgresql: res.postgresqlStatus, redis: res.redisStatus };
};

export const set_site_favicon = async (url: string): Promise<ISucceed> => {
  const res = await query<{
    setSiteFavicon: ISucceed;
  }>(
    `
mutation call($url: String!){
  setSiteFavicon(url: $url){
    createdAt
  }
}
`,
    { url }
  );
  return res.setSiteFavicon;
};
export const delete_site_gab_code = async (): Promise<ISucceed> => {
  const res = await query<{
    deleteSiteGabCode: ISucceed;
  }>(
    `
mutation call{
  deleteSiteGabCode{
    createdAt
  }
}
`,
    {}
  );
  return res.deleteSiteGabCode;
};
export const set_site_gab_code = async (
  code: string,
  name: string
): Promise<ISucceed> => {
  const res = await query<{
    setSiteGabCode: ISucceed;
  }>(
    `
mutation call($code: String!, $name: String!){
  setSiteGabCode(code: $code, name: $name){
    createdAt
  }
}
`,
    { code, name }
  );
  return res.setSiteGabCode;
};
export const delete_site_icp_code = async (): Promise<ISucceed> => {
  const res = await query<{
    deleteSiteIcpCode: ISucceed;
  }>(
    `
mutation call{
  deleteSiteIcpCode{
    createdAt
  }
}
`,
    {}
  );
  return res.deleteSiteIcpCode;
};
export const set_site_icp_code = async (code: string): Promise<ISucceed> => {
  const res = await query<{
    setSiteIcpCode: ISucceed;
  }>(
    `
mutation call($code: String!){
  setSiteIcpCode(code: $code){
    createdAt
  }
}
`,
    { code }
  );
  return res.setSiteIcpCode;
};
export const set_site_keywords = async (items: string[]): Promise<ISucceed> => {
  const res = await query<{
    setSiteKeywords: ISucceed;
  }>(
    `
mutation call($items: [String!]!){
  setSiteKeywords(items: $items){
    createdAt
  }
}
`,
    { items }
  );
  return res.setSiteKeywords;
};
export const set_site_authors = async (items: IAuthor[]): Promise<ISucceed> => {
  const res = await query<{
    setSiteAuthors: ISucceed;
  }>(
    `
mutation call($items: [SiteAuthorRequestItem!]!){
  setSiteAuthors(items: $items){
    createdAt
  }
}
`,
    { items }
  );
  return res.setSiteAuthors;
};
export const set_site_info = async (
  title: string,
  subhead: string,
  description: string,
  copyright: string
): Promise<ISucceed> => {
  const res = await query<{
    setSiteInfo: ISucceed;
  }>(
    `
mutation call($title: String!, $subhead: String!, $description: String!, $copyright: String!){
  setSiteInfo(title: $title, subhead: $subhead, description: $description, copyright: $copyright){
    createdAt
  }
}
`,
    { title, subhead, description, copyright }
  );
  return res.setSiteInfo;
};

interface IBaiduSiteVerification {
  content: string;
  code: string;
}
export const set_baidu_site_verification = async (
  code: string,
  content: string
): Promise<ISucceed> => {
  const res = await query<{
    setBaiduSiteVerification: ISucceed;
  }>(
    `
mutation call($code: String!, $content: String!){
  setBaiduSiteVerification(code: $code, content: $content){
    createdAt
  }
}
`,
    { content, code }
  );
  return res.setBaiduSiteVerification;
};
export const delete_baidu_site_verification = async (): Promise<ISucceed> => {
  const res = await query<{
    deleteBaiduSiteVerification: ISucceed;
  }>(
    `
mutation call{
  deleteBaiduSiteVerification{
    createdAt
  }
}
`,
    {}
  );
  return res.deleteBaiduSiteVerification;
};
export const get_baidu_site_verification =
  async (): Promise<IBaiduSiteVerification> => {
    const res = await query<{
      getBaiduSiteVerification: IBaiduSiteVerification;
    }>(
      `
query call{
  getBaiduSiteVerification{
    content, code
  }
}
`,
      {}
    );
    return res.getBaiduSiteVerification;
  };

interface IGoogleSiteVerification {
  code: string;
}
export const set_google_site_verification = async (
  code: string
): Promise<ISucceed> => {
  const res = await query<{
    setGoogleSiteVerification: ISucceed;
  }>(
    `
mutation call($code: String!){
  setGoogleSiteVerification(code: $code){
    createdAt
  }
}
`,
    { code }
  );
  return res.setGoogleSiteVerification;
};
export const delete_google_site_verification = async (): Promise<ISucceed> => {
  const res = await query<{
    deleteGoogleSiteVerification: ISucceed;
  }>(
    `
mutation call{
  deleteGoogleSiteVerification{
    createdAt
  }
}
`,
    {}
  );
  return res.deleteGoogleSiteVerification;
};
export const get_google_site_verification =
  async (): Promise<IGoogleSiteVerification> => {
    const res = await query<{
      getGoogleSiteVerification: IGoogleSiteVerification;
    }>(
      `
query call{
  getGoogleSiteVerification{
    code
  }
}
`,
      {}
    );
    return res.getGoogleSiteVerification;
  };

interface IGoogleRecaptcha {
  siteKey: string;
  secret: string;
}
export const set_google_recaptcha = async (
  siteKey: string,
  secret: string
): Promise<ISucceed> => {
  const res = await query<{
    setGoogleRecaptcha: ISucceed;
  }>(
    `
  mutation call($siteKey: String!, $secret: String!){
    setGoogleRecaptcha(siteKey: $siteKey, secret: $secret){
      createdAt
    }
  }
  `,
    { siteKey, secret }
  );
  return res.setGoogleRecaptcha;
};
export const delete_google_recaptcha = async (): Promise<ISucceed> => {
  const res = await query<{
    deleteGoogleRecaptcha: ISucceed;
  }>(
    `
  mutation call{
    deleteGoogleRecaptcha{
      createdAt
    }
  }
  `,
    {}
  );
  return res.deleteGoogleRecaptcha;
};
export const get_google_recaptcha = async (): Promise<IGoogleRecaptcha> => {
  const res = await query<{
    getGoogleRecaptcha: IGoogleRecaptcha;
  }>(
    `
  query call{
    getGoogleRecaptcha{
      siteKey, secret
    }
  }
  `,
    {}
  );
  return res.getGoogleRecaptcha;
};

interface IIndexNowSiteVerification {
  key: string;
}
export const set_index_now_site_verification = async (
  key: string
): Promise<ISucceed> => {
  const res = await query<{
    setIndexNowSiteVerification: ISucceed;
  }>(
    `
  mutation call($key: String!){
    setIndexNowSiteVerification(key: $key){
      createdAt
    }
  }
  `,
    { key }
  );
  return res.setIndexNowSiteVerification;
};
export const delete_index_now_site_verification =
  async (): Promise<ISucceed> => {
    const res = await query<{
      deleteIndexNowSiteVerification: ISucceed;
    }>(
      `
  mutation call{
    deleteIndexNowSiteVerification{
      createdAt
    }
  }
  `,
      {}
    );
    return res.deleteIndexNowSiteVerification;
  };
export const get_index_now_site_verification =
  async (): Promise<IIndexNowSiteVerification> => {
    const res = await query<{
      getIndexNowSiteVerification: IIndexNowSiteVerification;
    }>(
      `
  query call{
    getIndexNowSiteVerification{
      key
    }
  }
  `,
      {}
    );
    return res.getIndexNowSiteVerification;
  };

File: tuberose/lib/tuberose_web/schema/resolvers/site.ex
defmodule TuberoseWeb.Resolvers.Site do
  import Ecto.Query
  require Logger

  def delete_index_now_site_verification(_parent, _args, %{context: context}) do
    unless Tuberose.Atropa.Client.administrator?(context.current_user.id) do
      raise ArgumentError, message: "Forbidden"
    end

    Tuberose.Repo.transaction(fn ->
      it =
        from(p in Tuberose.Setting,
          where: is_nil(p.user_id) and p.key == "index-now.site-verification"
        )
        |> first
        |> Tuberose.Repo.one()

      if it do
        Tuberose.Repo.delete(it)
      end
    end)

    {:ok, %{created_at: DateTime.utc_now()}}
  end

  def get_index_now_site_verification(_parent, _args, %{context: context}) do
    unless Tuberose.Atropa.Client.administrator?(context.current_user.id) do
      raise ArgumentError, message: "Forbidden"
    end

    item = Tuberose.KV.get("index-now.site-verification")

    if item,
      do: {:ok, %{key: item.key}},
      else: {:ok, %{key: ""}}
  end

  def set_index_now_site_verification(_parent, %{key: key}, %{
        context: context
      }) do
    unless Tuberose.Atropa.Client.administrator?(context.current_user.id) do
      raise ArgumentError, message: "Forbidden"
    end

    Tuberose.Repo.transaction(fn ->
      Tuberose.KV.set("index-now.site-verification", %{key: key}, true)
    end)

    {:ok, %{created_at: DateTime.utc_now()}}
  end

  def delete_baidu_site_verification(_parent, _args, %{context: context}) do
    unless Tuberose.Atropa.Client.administrator?(context.current_user.id) do
      raise ArgumentError, message: "Forbidden"
    end

    Tuberose.Repo.transaction(fn ->
      it =
        from(p in Tuberose.Setting,
          where: is_nil(p.user_id) and p.key == "baidu.site-verification"
        )
        |> first
        |> Tuberose.Repo.one()

      if it do
        Tuberose.Repo.delete(it)
      end
    end)

    {:ok, %{created_at: DateTime.utc_now()}}
  end

  def get_baidu_site_verification(_parent, _args, %{context: context}) do
    unless Tuberose.Atropa.Client.administrator?(context.current_user.id) do
      raise ArgumentError, message: "Forbidden"
    end

    item = Tuberose.KV.get("baidu.site-verification")

    if item,
      do: {:ok, %{code: item.code, content: item.content}},
      else: {:ok, %{code: "", content: ""}}
  end

  def set_baidu_site_verification(_parent, %{code: code, content: content}, %{
        context: context
      }) do
    unless Tuberose.Atropa.Client.administrator?(context.current_user.id) do
      raise ArgumentError, message: "Forbidden"
    end

    Tuberose.Repo.transaction(fn ->
      Tuberose.KV.set("baidu.site-verification", %{code: code, content: content}, true)
    end)

    {:ok, %{created_at: DateTime.utc_now()}}
  end

  def delete_google_site_verification(_parent, _args, %{context: context}) do
    unless Tuberose.Atropa.Client.administrator?(context.current_user.id) do
      raise ArgumentError, message: "Forbidden"
    end

    Tuberose.Repo.transaction(fn ->
      it =
        from(p in Tuberose.Setting,
          where: is_nil(p.user_id) and p.key == "google.site-verification"
        )
        |> first
        |> Tuberose.Repo.one()

      if it do
        Tuberose.Repo.delete(it)
      end
    end)

    {:ok, %{created_at: DateTime.utc_now()}}
  end

  def get_google_site_verification(_parent, _args, %{context: context}) do
    unless Tuberose.Atropa.Client.administrator?(context.current_user.id) do
      raise ArgumentError, message: "Forbidden"
    end

    item = Tuberose.KV.get("google.site-verification")

    if item,
      do: {:ok, %{code: item.code}},
      else: {:ok, %{code: ""}}
  end

  def set_google_site_verification(_parent, %{code: code}, %{
        context: context
      }) do
    unless Tuberose.Atropa.Client.administrator?(context.current_user.id) do
      raise ArgumentError, message: "Forbidden"
    end

    Tuberose.Repo.transaction(fn ->
      Tuberose.KV.set("google.site-verification", %{code: code}, true)
    end)

    {:ok, %{created_at: DateTime.utc_now()}}
  end

  def delete_google_recaptcha(_parent, _args, %{context: context}) do
    unless Tuberose.Atropa.Client.administrator?(context.current_user.id) do
      raise ArgumentError, message: "Forbidden"
    end

    Tuberose.Repo.transaction(fn ->
      it =
        from(p in Tuberose.Setting,
          where: is_nil(p.user_id) and p.key == "google.recaptcha"
        )
        |> first
        |> Tuberose.Repo.one()

      if it do
        Tuberose.Repo.delete(it)
      end
    end)

    {:ok, %{created_at: DateTime.utc_now()}}
  end

  def get_google_recaptcha(_parent, _args, %{context: context}) do
    unless Tuberose.Atropa.Client.administrator?(context.current_user.id) do
      raise ArgumentError, message: "Forbidden"
    end

    item = Tuberose.KV.get("google.recaptcha")

    if item,
      do: {:ok, %{site_key: item.site_key, secret: item.secret}},
      else: {:ok, %{site_key: "", secret: ""}}
  end

  def set_google_recaptcha(_parent, %{site_key: site_key, secret: secret}, %{context: context}) do
    unless Tuberose.Atropa.Client.administrator?(context.current_user.id) do
      raise ArgumentError, message: "Forbidden"
    end

    Tuberose.Repo.transaction(fn ->
      Tuberose.KV.set("google.recaptcha", %{site_key: site_key, secret: secret}, true)
    end)

    {:ok, %{created_at: DateTime.utc_now()}}
  end

  def routes(_parent, _args, %{context: context}) do
    locale = context.locale
    items = []

    personal =
      if context.current_user do
        [
          %{
            path: "/dashboard/personal",
            name: Tuberose.I18N.t(locale, "personal.index.title"),
            children: [
              %{
                path: "/dashboard/personal/logs",
                name: Tuberose.I18N.t(locale, "personal.logs.title"),
                children: []
              },
              %{
                path: "/dashboard/personal/profile",
                name: Tuberose.I18N.t(locale, "personal.profile.title"),
                children: []
              },
              %{
                path: "/dashboard/personal/attachments",
                name: Tuberose.I18N.t(locale, "personal.attachments.title"),
                children: []
              }
            ]
          }
        ]
      else
        []
      end

    settings =
      if context.current_user != nil and
           Tuberose.Atropa.Client.administrator?(context.current_user.id) do
        [
          %{
            path: "/dashboard/settings",
            name: Tuberose.I18N.t(locale, "settings.index.title"),
            children: [
              %{
                path: "/dashboard/settings/site/status",
                name: Tuberose.I18N.t(locale, "settings.site.status.title"),
                children: []
              },
              %{
                path: "/dashboard/settings/site/info",
                name: Tuberose.I18N.t(locale, "settings.site.info.title"),
                children: []
              },
              %{
                path: "/dashboard/settings/site/seo",
                name: Tuberose.I18N.t(locale, "settings.site.seo.title"),
                children: []
              },
              %{
                path: "/dashboard/settings/locales",
                name: Tuberose.I18N.t(locale, "settings.locales.index.title"),
                children: []
              }
            ]
          }
        ]
      else
        []
      end

    {:ok, items ++ personal ++ settings}
  end

  def layout(_parent, _args, %{context: context}) do
    {:ok,
     %{
       favicon: Tuberose.KV.get("site.author") || "/my/favicon.svg",
       title: Tuberose.I18N.t(context.locale, "site.title"),
       subhead: Tuberose.I18N.t(context.locale, "site.subhead"),
       description: Tuberose.I18N.t(context.locale, "site.description"),
       copyright: Tuberose.I18N.t(context.locale, "site.copyright"),
       author: Tuberose.KV.get("site.author"),
       keywords: Tuberose.KV.get("site.keywords") || [],
       icp: Tuberose.KV.get("site.icp"),
       gab: Tuberose.KV.get("site.gab"),
       locale: context.locale,
       languages: Gettext.known_locales(TuberoseWeb.Gettext)
     }}
  end
end
